flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,87 +75,360 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
+        /*num_teams_upper=*/nullptr, /*if_expr=*/nullptr,
+        /*thread_limit=*/nullptr,
+        /*allocate_vars=*/llvm::SmallVector<mlir::Value>{},
+        /*allocators_vars=*/llvm::SmallVector<mlir::Value>{},
+        /*reduction_vars=*/llvm::SmallVector<mlir::Value>{},
+        /*reductions=*/nullptr);
+
+    mlir::Block *teamsBlock = rewriter.createBlock(&teamsOp.getRegion());
+    rewriter.create<mlir::omp::TerminatorOp>(loc);
+    rewriter.setInsertionPointToStart(teamsBlock);
+
+    return teamsOp;
+  }
+
+  mlir::omp::DistributeOp
+  genDistributeOp(mlir::Location loc,
+                  mlir::ConversionPatternRewriter &rewriter) const {
+    mlir::omp::DistributeOp distOp = rewriter.create<mlir::omp::DistributeOp>(
+        loc, /*dist_schedule_static=*/nullptr,
+        /*chunk_size=*/nullptr,
+        /*allocate_vars=*/mlir::ValueRange(),
+        /*allocators_vars=*/mlir::ValueRange(),
+        /*order_val=*/nullptr);
+
+    mlir::Block *distBlock = rewriter.createBlock(&distOp.getRegion());
+    rewriter.create<mlir::omp::TerminatorOp>(loc);
+    rewriter.setInsertionPointToStart(distBlock);
+
+    return distOp;
+  }
+
+  mlir::omp::ParallelOp
+  genParallelOp(mlir::Location loc, mlir::ConversionPatternRewriter &rewriter,
+                const llvm::SmallVector<mlir::Value> &liveIns,
+                mlir::IRMapping &mapper) const {
+    mlir::omp::ParallelOp parallelOp =
+        rewriter.create<mlir::omp::ParallelOp>(loc);
+    mlir::Block *parRegion = rewriter.createBlock(&parallelOp.getRegion());
+    rewriter.create<mlir::omp::TerminatorOp>(loc);
+    rewriter.setInsertionPointToStart(parRegion);
+
+    mlir::Operation *indVarMemDef = liveIns.front().getDefiningOp();
+
+    assert(
+        indVarMemDef != nullptr &&
+        "Induction variable memdef is expected to have a defining operation.");
+
+    llvm::SmallSetVector<mlir::Operation *, 2> indVarDeclareAndAlloc;
+    for (auto operand : indVarMemDef->getOperands())
+      indVarDeclareAndAlloc.insert(operand.getDefiningOp());
+    indVarDeclareAndAlloc.insert(indVarMemDef);
+
+    for (mlir::Operation *opToClone : indVarDeclareAndAlloc)
+      rewriter.clone(*opToClone, mapper);
+
+    return parallelOp;
+  }
+
+  mlir::omp::WsloopOp genWsLoopOp(mlir::ConversionPatternRewriter &rewriter,
+                                  fir::DoLoopOp doLoop,
+                                  mlir::IRMapping &mapper) const {
+    mlir::Operation *lbOp = doLoop.getLowerBound().getDefiningOp();
+    mlir::Operation *ubOp = doLoop.getUpperBound().getDefiningOp();
+    mlir::Operation *stepOp = doLoop.getStep().getDefiningOp();
+
+    // Clones the chain of ops defining a certain loop bound or its step into
+    // the parallel region. For example, if the value of a bound is defined by a
+    // `fir.convert`op, this lambda clones the `fir.convert` as well as the
+    // value it converts from. We do this since `omp.target` regions are
+    // isolated from above.
```

Comment:
I think it makes sense to copy the ops related to constants and type casts into the target region (rather than just before `omp.wsloop`, since that means introducing operations between the leaf constructs of a composite, which will be disallowed by loop wrappers), but shouldn't everything be already mapped and present as an entry block argument to `omp.target` by this time, due to it being detected as a live-in value?

Reply:
I think one issue that may arise, is that map operations only take pointer-like values, Index values are not mappable (you can circumvent this by creating the right operations to load it into memory (so convert it to a ref) and pass that instead I would imagine), there was some contention with this when @TIFitis tried to do it originally, so we rolled back to just using pointer-like arguments in map operations.

Reply:
> but shouldn't everything be already mapped and present as an entry block argument to omp.target by this time, due to it being detected as a live-in value?

For the simple cases I tested so far, this is not the situation since LB, UB, step and constants; so they are not detected as live-ins. Only the induction variable is a live-in.

> I think it makes sense to copy the ops related to constants and type casts into the target region (rather than just before omp.wsloop,

If you don't mind, I will leave that until we merge your work on composite constructs from upstream.